### DISCUSSION TITLE
Sampler1DReal to get displaced qp values, bug?

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#22935] [Q&A Modules: General] [answered]
I've got a two-element thermal expansion problem with a temperature gradient along x. I want to use the `Sampler1DReal` VPP to get the temperature, which is stored in an AuxVariable, at the _displaced_ quadrature points. I'm doing this using an intermediate `ParsedMaterial` that is just the value of the temperature AuxVariable, since the VPP only accepts materials. 

The temperature values are as expected for the VPP, but when I set `use_displaced_mesh=true` in the VPP, the temperature values are wrong (though the coordinates are as expected). Seems like the temperature values for the first element are just repeats of those the second element? Am I using the VPP incorrectly? (It says `1D` in the name after all)

Input file (sorry its so long)
```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 2
  ny = 1
[]

[GlobalParams]
  displacements = 'disp_x disp_y'
[]

[AuxVariables]
  [./temp]
    order = first
    family = lagrange
  [../]
[]

[Functions]
  [./temperature_load]
    type = ParsedFunction
    value = t*(500.0)+300.0+100*x
  [../]
[]

[Modules]
  [./TensorMechanics]
    [./Master]
      [./all]
        strain = finite
        incremental = true
        add_variables = true
        eigenstrain_names = eigenstrain
      [../]
    [../]
  [../]
[]

[AuxKernels]
  [./tempfuncaux]
    type = FunctionAux
    variable = temp
    function = temperature_load
  [../]
[]

[BCs]
  [./x_bot]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0.0
  [../]
  [./y_bot]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0.0
  [../]
[]

[Materials]
  [./elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 2.1e5
    poissons_ratio = 0.3
  [../]
  [./small_stress]
    type = ComputeFiniteStrainElasticStress
  [../]
  [./thermal_expansion_strain]
    type = ComputeThermalExpansionEigenstrain
    stress_free_temperature = 298
    thermal_expansion_coeff = 1.3e-5
    temperature = temp
    eigenstrain_name = eigenstrain
  [../]
  [mat]
    type = ParsedMaterial
    function = temp
    args = temp
    f_name = temperature_material
  []
[]

[Executioner]
  type = Transient
  solve_type = 'PJFNK'

  l_max_its = 50
  nl_max_its = 50
  nl_rel_tol = 1e-12
  nl_abs_tol = 1e-10
  l_tol = 1e-9

  dt = 0.0125
  num_steps = 1
[]

[Outputs]
  exodus = true
  csv = true
[]

[Postprocessors]
  [./avg_temperature]
    type = AverageNodalVariableValue
    variable = temp
  [../]
  [./max_temperature]
    type = NodalExtremeValue
    variable = temp
    value_type = max
  []
[]

[VectorPostprocessors]
  [sampler1dreal]
    type = Sampler1DReal
    block = 0
    property = temperature_material
    sort_by = x
  []
  [sampler1dreal_disp]
    type = Sampler1DReal
    block = 0
    property = temperature_material
    sort_by = x
    use_displaced_mesh = true
  []
[]

```

VPP output:

| id | temperature_material | x | y | z |
| -- | -------------------- | - | - | - |
| 0.10566243270259 | **316.81624327026** | 0.10566243270259 | 0.21132486540519 | 0 |
| 0.10566243270259 | **316.81624327026** | 0.10566243270259 | 0.78867513459481 | 0 |
| 0.39433756729741 | **345.68375672974** | 0.39433756729741 | 0.21132486540519 | 0 |
| 0.39433756729741 | **345.68375672974** | 0.39433756729741 | 0.78867513459481 | 0 |
| 0.60566243270259 | **366.81624327026** | 0.60566243270259 | 0.21132486540519 | 0 |
| 0.60566243270259 | **366.81624327026** | 0.60566243270259 | 0.78867513459481 | 0 |
| 0.89433756729741 | **395.68375672974** | 0.89433756729741 | 0.21132486540519 | 0 |
| 0.89433756729741 | **395.68375672974** | 0.89433756729741 | 0.78867513459481 | 0 |

Displaced VPP output:

| id | temperature_material | x | y | z |
| -- | -------------------- | - | - | - |
| 0.1056958406132 | **366.81624327026** | 0.1056958406132 | 0.78914481344896 | 0 |
| 0.10573288589375 | **366.81624327026** | 0.10573288589375 | 0.21145071547486 | 0 |
| 0.39446224731715 | **395.68375672974** | 0.39446224731715 | 0.78935207603834 | 0 |
| 0.39460050218634 | **395.68375672974** | 0.39460050218634 | 0.2115062513183 | 0 |
| 0.60597117492747 | **366.81624327026** | 0.60597117492747 | 0.78952409424419 | 0 |
| 0.60615762034783 | **366.81624327026** | 0.60615762034783 | 0.21155234345764 | 0 |
| 0.89505790470932 | **395.68375672974** | 0.89505790470932 | 0.78978679413302 | 0 |
| 0.89527479957529 | **395.68375672974** | 0.89527479957529 | 0.21162273368071 | 0 |

**Comment by GiudGiud** [comment]
Hello

The Sampler1DReal was written for 1D thermal hydraulics, which does not use displaced meshes afaik.
The code is not considering the displaced mesh or problem

The q_points are correct because use_displaced_mesh makes sure of that. 

Adapting it to use displaced mesh and problem would not be too difficult if you want to give it a try

Guillaume

**Reply by jessecarterMOOSE** [reply]
You have much more confidence in my ability than I do. Can I get a hint? I think this could be a useful capability.

**Reply by GiudGiud** [reply]
anywhere where it says problem you use getDisplacedProblem() from the FEProblem instead
https://mooseframework.inl.gov/docs/doxygen/moose/classFEProblemBase.html#a6bdd4ae4d1d41245e59c8bccf35e3f94

and if you get the mesh from the displaced problem it ll be the displaced mesh

**Comment by friedmud** [comment, marked as answer]
The "displaced" option for PPs/VPPs works somewhat counter-intuitively:

1.  With `displaced_mesh = true` the _mesh_ will move across the _stationary_ PP/VPP. (Think Lagrangian motion of a mesh moving across/through something

2.  When it's false the PP stays in the elements you put it in in the beginning (because they are not moving).  So it will effectively "move" with the material (because even though the displaced mesh is moving - the PP/VPP will always sample the same positions)

Just keep in mind that it's the _mesh_ that is displaced - not the PP/VPP

**Reply by jessecarterMOOSE** [reply]
Interesting. Didn't realize that. So if you're sampling a point in space, and the mesh displaces to the _right_, the coordinates returned by the VPP will shift to the _left_ because the point stayed still but the mesh moved?

**Reply by friedmud** [reply]
Exactly.

Well - the effect is like that.  But the sampling point stays still while the mesh moves across it.

(Just being extra clear incase others find this and read along)

**Comment by jessecarterMOOSE** [comment]
What's the difference between `_fe_problem` and `_subproblem`?

**Reply by GiudGiud** [reply]
Subproblem is an intermediate class between Problem and FEProblemBase

I think the idea was to split routines between several levels of inheritance, as these classes are huge
It s also to hold some routines that may be shared between multiple types of problems. 

For your case, they point to "same object" but with different class types. You get more routines available on the _subproblem

@lindsayad or @friedmud in case I m wrong

**Reply by lindsayad** [reply]
If `use_displaced_mesh = true`, then `_subproblem` points to the displaced problem, not the finite element problem (`_fe_problem`). `_subproblem` exists because objects will either be on the undisplaced subproblem (the finite element problem) or the displaced subproblem
